tgui/packages/tgui/components/Box.js

```diff
@@ -0,0 +1,226 @@
+  },
+  opacity: mapRawPropTo('opacity'),
+  textAlign: mapRawPropTo('text-align'),
+  verticalAlign: mapRawPropTo('vertical-align'),
+  // Boolean props
+  inline: mapBooleanPropTo('display', 'inline-block'),
+  bold: mapBooleanPropTo('font-weight', 'bold'),
+  italic: mapBooleanPropTo('font-style', 'italic'),
+  nowrap: mapBooleanPropTo('white-space', 'nowrap'),
+  // Margins
+  m: mapDirectionalUnitPropTo('margin', halfUnit, [
+    'top', 'bottom', 'left', 'right',
+  ]),
+  mx: mapDirectionalUnitPropTo('margin', halfUnit, [
+    'left', 'right',
+  ]),
+  my: mapDirectionalUnitPropTo('margin', halfUnit, [
+    'top', 'bottom',
+  ]),
+  mt: mapUnitPropTo('margin-top', halfUnit),
+  mb: mapUnitPropTo('margin-bottom', halfUnit),
+  ml: mapUnitPropTo('margin-left', halfUnit),
+  mr: mapUnitPropTo('margin-right', halfUnit),
+  // Margins
+  p: mapDirectionalUnitPropTo('padding', halfUnit, [
+    'top', 'bottom', 'left', 'right',
+  ]),
+  px: mapDirectionalUnitPropTo('padding', halfUnit, [
+    'left', 'right',
+  ]),
+  py: mapDirectionalUnitPropTo('padding', halfUnit, [
+    'top', 'bottom',
+  ]),
+  pt: mapUnitPropTo('padding-top', halfUnit),
+  pb: mapUnitPropTo('padding-bottom', halfUnit),
+  pl: mapUnitPropTo('padding-left', halfUnit),
+  pr: mapUnitPropTo('padding-right', halfUnit),
+  // Color props
+  color: mapColorPropTo('color'),
+  textColor: mapColorPropTo('color'),
+  backgroundColor: mapColorPropTo('background-color'),
+  // Utility props
+  fillPositionedParent: (style, value) => {
+    if (value) {
+      style['position'] = 'absolute';
+      style['top'] = 0;
+      style['bottom'] = 0;
+      style['left'] = 0;
+      style['right'] = 0;
+    }
+  },
+};
+
+export const computeBoxProps = props => {
+  const computedProps = {};
+  const computedStyles = {};
+  // Compute props
+  for (let propName of Object.keys(props)) {
+    if (propName === 'style') {
+      continue;
+    }
+    const propValue = props[propName];
+    const mapPropToStyle = styleMapperByPropName[propName];
+    if (mapPropToStyle) {
+      mapPropToStyle(computedStyles, propValue);
+    }
+    else {
+      computedProps[propName] = propValue;
+    }
+  }
+  // Concatenate styles
+  let style = '';
+  for (let attrName of Object.keys(computedStyles)) {
+    const attrValue = computedStyles[attrName];
+    style += attrName + ':' + attrValue + ';';
+  }
+  if (props.style) {
+    for (let attrName of Object.keys(props.style)) {
+      const attrValue = props.style[attrName];
+      style += attrName + ':' + attrValue + ';';
+    }
+  }
+  if (style.length > 0) {
+    computedProps.style = style;
+  }
+  return computedProps;
+};
+
+export const computeBoxClassName = props => {
+  const color = props.textColor || props.color;
+  const backgroundColor = props.backgroundColor;
+  return classes([
+    isColorClass(color) && 'color-' + color,
+    isColorClass(backgroundColor) && 'color-bg-' + backgroundColor,
+  ]);
+};
+
+export const Box = props => {
```

Comment:
Yes, and I regret adding that many properties on it.

Reply:
Given its significant usage as an underlying thing in many other components I'm not sure that it could be decoupled very easily, but this is one of those bits that I'll be looking at from a "how do we make this both easy to use and easy to use in a standard way?" approach.

Not requiring for this PR, but I think we might want to look at some "preconfigured" Box variants that do the relevant styling and are imported into the various other places they're used. That way people can easily use a "standard" Box type, but can still import the current (unbelievably flexible) Box if they need to do something really funky.